src/generate/rust.rs

```diff
@@ -43,237 +44,233 @@ struct ParseNode {
     ty: Option<Src>,
 }
 
-struct Variant<'a, Pat> {
-    rule: Rc<Rule<Pat>>,
-    name: &'a str,
-    fields: OrderMap<&'a str, OrderSet<Vec<usize>>>,
+struct Variant {
+    rule: IRule,
+    name: IStr,
+    fields: IndexMap<IStr, FieldPathset>,
 }
 
 trait RuleWithNamedFieldsMethods<Pat> {
-    fn find_variant_fields(&self) -> Option<Vec<Variant<'_, Pat>>>;
+    fn find_variant_fields(&self, cx: &mut Context<Pat>) -> Option<Vec<Variant>>;
 }
 
-impl<Pat: PartialEq> RuleWithNamedFieldsMethods<Pat> for RuleWithNamedFields<Pat> {
-    fn find_variant_fields(&self) -> Option<Vec<Variant<'_, Pat>>> {
-        if let Rule::Or(cases) = &*self.rule {
+impl<Pat> RuleWithNamedFieldsMethods<Pat> for RuleWithNamedFields {
+    fn find_variant_fields(&self, cx: &mut Context<Pat>) -> Option<Vec<Variant>> {
+        if let Rule::Or(cases) = &cx[self.rule] {
             if self.fields.is_empty() {
                 return None;
             }
-            let mut variants: Vec<_> = cases
-                .iter()
-                .map(|rule| Variant {
-                    rule: rule.clone(),
-                    name: "",
-                    fields: OrderMap::new(),
-                })
-                .collect();
-            for (field, paths) in &self.fields {
-                for path in paths {
+
+            let mut variant_names: Vec<_> = vec![None; cases.len()];
+            let mut variant_fields: Vec<_> = vec![IndexMap::new(); cases.len()];
+            for (&field, paths) in &self.fields {
+                for path in &paths.0 {
                     match path[..] {
                         [] => return None,
-                        [variant] if variants[variant].name != "" => return None,
-                        [variant] => variants[variant].name = field,
+                        [variant] => {
+                            if variant_names[variant].is_some() {
```

Comment:
Why not continue to use a guard here?

Reply:
I think I wanted to avoid accessing `variants` twice, but I ended up making 2 separate `Vec`s.